web/src/reload.ts

```diff
@@ -200,76 +203,137 @@ export function initiate({
+            }
+            function compose_started_handler(): void {
+                // If the user stops being idle and starts composing a
+                // message, switch to the compose-open timeouts.
+                idle_control.cancel();
+                idle_control = $(document).idle({
+                    idle: composing_idle_timeout,
+                    onIdle: reload_from_idle,
+                });
+                $(document).off("compose_started.zulip", compose_started_handler);
+                $(document).on(
+                    "compose_canceled.zulip compose_finished.zulip",
+                    compose_done_handler,
+                );
+            }
+
+            if (compose_state.composing()) {
+                idle_control = $(document).idle({
+                    idle: composing_idle_timeout,
+                    onIdle: reload_from_idle,
+                });
+                $(document).on(
+                    "compose_canceled.zulip compose_finished.zulip",
+                    compose_done_handler,
+                );
+            } else {
+                idle_control = $(document).idle({
+                    idle: basic_idle_timeout,
+                    onIdle: reload_from_idle,
+                });
+                $(document).on("compose_started.zulip", compose_started_handler);
+            }
+        },
+        error(xhr) {
+            // We need to respect the server's rate-limiting headers, but beyond
+            // that, we also want to avoid contributing to a thundering herd if
+            // the server is giving us 500s/502s.
+            //
+            // So we do the maximum of the retry-after header and an exponential
+            // backoff with ratio sqrt(2) and half jitter. Starts at 1-2s and ends at
+            // 45-90s after enough failures.
+            server_settings_failures += 1;
+            const backoff_scale = Math.min(2 ** ((server_settings_failures + 1) / 2), 90);
+            const backoff_delay_secs = ((1 + Math.random()) / 2) * backoff_scale;
+            let rate_limit_delay_secs = 0;
+            const rate_limited_error_schema = z.object({
+                "retry-after": z.number(),
+                code: z.literal("RATE_LIMIT_HIT"),
+            });
+            const parsed = rate_limited_error_schema.safeParse(xhr.responseJSON);
+            if (xhr.status === 429 && parsed?.success && parsed?.data) {
+                // Add a bit of jitter to the required delay suggested by the
+                // server, because we may be racing with other copies of the web
+                // app.
+                rate_limit_delay_secs = parsed.data["retry-after"] + Math.random() * 0.5;
+            }
+
+            const retry_delay_secs = Math.max(backoff_delay_secs, rate_limit_delay_secs);
+            setTimeout(() => {
+                initiate({immediate, save_compose, send_after_reload, message_html});
+            }, retry_delay_secs * 1000);
```

Comment:
This is basically the same as what's used in `get_events` for the retry delay.

Reply:
OK. This may be something we find reasonable to extract as a `util.get_retry_backoff(xhr, failures) type of` function that returns a simple object with any state properties to save as well as the desired delay.

Reply:
If you end up doing another pass on this, let's do it; it's fairly likely we'll end up regretting forgetting to.

Reply:
Since I did another pass to fix the test failure, I added a commit to extract the delay backoff algorithms into the suggested helper function, but not the other state properties.